### ISSUE TITLE
Should().Be() / NotBe() prefers IComparable<T>.Compare over object.Equals

### ISSUE THREAD

**Issue by ulrichb** [fluentassertions/fluentassertions#172]
Given the following class with _value semantics_.

``` C#
public class Ranked<T> : IComparable<Ranked<T>>
{
    public Ranked (int ranking, T name) { Ranking = ranking; Item = name; }

    public int Ranking { get; private set; }
    public T Item { get; private set; }

    public override bool Equals (object obj)
    {
        if (ReferenceEquals (null, obj))
            return false;
        if (obj.GetType() != this.GetType())
            return false;
        var other = (Ranked<T>) obj;
        return Ranking == other.Ranking && Equals (Item, other.Item);
    }

    public override int GetHashCode () { /* ... */ }

    public int CompareTo (Ranked<T> other)
    {
        if (other == null) return 1;

        return this.Ranking.CompareTo (other.Ranking);
    }
}
```

At the moment (v3.2.1) the assertion `new Ranked<string>(1, "Some item").Should().Be (new Ranked<string>(1, "Other item"))` passes whereas casting the first part to object, `((object)new Ranked<string>(1, "Some item")).Should().Be (new Ranked<string>(1, "Other item"))`, fails. The first assertion calls `CompareTo`, which just compares the `Ranking` value and the second one calls `Equals`.

The reason for the different behavior of the two assertions is the `IComparable<>` overload of `Should()`. The idea of this overload is to be able to offer fluent methods for `IComparable<>` values, like `BeInRange()`. Unfortunately also `Be()` and `NotBe()` are "overridden" to check for `CompareTo() == 0` resp. `CompareTo() != 0`.

In my opinion this is not a good idea, because `Be()` / `NotBe()` are the _standard methods_ in FluentAssertions to check for _equality / inequality_. And sort ordering is not the same as equality. The example above shows a legitimate case, where just a subset of the properties is part of the ordering.

Thinking of another example of a `Person`-_entity_ where equality is defined by ID-equality (e.g. a `SocialSecurityNumber` property) and the ordering should be applied for `LastName / GivenName` shows that the set of properties used for equality / ordering don't even have to intersect.

We should think about a (breaking) change of the behavior of `ComparableTypeAssertions<T>.Be()`. So that `Should().Be()` always checks for equality and introduce a new `Should().HaveSameOrderingPositionAs()` (probably there's better name), which checks for `CompareTo() == 0`.

The reason why I would even prefer the breaking change is that a) it's unintuitive that `Should.Be()` delegates to `CompareTo` depending on the _compile time_ type of the left operand and b) the duality of `Should().Be()` is unsafe for example when adding `IComparable<T>` to the `Ranked<T>` example above _after_ there are already a lot of unit tests, which use `Should.Be()` and rely on its behavior.

Further notes:
- The MSDN documentation team strengthened the definition of the zero result of `IComparable.CompareTo` from to "This instance is equal to obj" to "This current instance occurs in the _same position in the sort order_ as the object specified by the CompareTo method" in .NET 3.5 (obviously to distinguish between equality and sort ordering). ~~Interestingly they didn't strengthen the definition of the generic version, but the semantics are the same for both interfaces.~~ (This [happened in the meantime](https://docs.microsoft.com/en-us/dotnet/api/system.icomparable-1#remarks).)
- NUnit's `Assert.AreEqual` and `Assert.That` only use `IEquatable<>.Equals` and `object.Equals` (depending on the runtime type), which both have the same semantics.
- MSTest's `Assert.AreEqual` only uses `object.Equals`.

**Comment by somewhatabstract** [comment]
Seems like changing `Be` and `NotBe` in this circumstance would be a breaking change for anyone who is relying on the current "broken" behavior.

**Comment by ulrichb** [comment]
@dennisdoomen What about adding this to #463?

**Comment by dennisdoomen** [comment]
> What about adding this to #463?

Might not be such a bad idea.

**Comment by ulrichb** [comment]
OK, cool. Could you also add it to the issue descripion?

**Comment by rubenrorije** [comment]
I just ran into this same issue (#1150). I understand that for backward compatibility reasons it is is difficult to change, but I was quite surprised that my equality test suddenly failed after implementing `IComparable<T>` on my class under test.

I fully agree with @ulrichb  (and ran into the exact problem described) 

> `Should().Be()` is unsafe for example when adding `IComparable<T>` to the `Ranked<T>` example above after there are already a lot of unit tests, which use `Should.Be()` and rely on it's behavior.

**Comment by rubenrorije** [comment]
Since this issue is open for quite a while now, would it be a solution for the time being to change the implementation of `Be` in `ComparableTypeAssertions` to also check equality? 
It would be a breaking change, but would only result in more assertions and would catch more errors that previously would be unnoticed. Note that it assumes that equality is implemented when implementing `IComparable`, but this seems like a safe assumption to me. 

It would not solve my problem above, but I am more concerned about my and other developer's incorrect expactations than my issue above, that is easily circumvented.

**Comment by jnyrup** [comment]
5 years later, this was fixed in #1177.
